xmlserializer/XmlUtil.cpp

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2015, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libxml/parser.h>
+
+/* Provides miscellaneous utility functions */
+class CXmlUtil
+{
+public:
+    /* xmlInitParser must be called in the main thread if multithreading is required
+     * @see http://xmlsoft.org/threads.html
+     * @see http://www.programmershare.com/1669782/
+     */
+    CXmlUtil() {
+        xmlInitParser();
+    }
+};
+
+static CXmlUtil xml_init;
```

Comment:
- xml_init->xmlInit
  - Do you mean that there is a race if some other code in the process 
call xmlInitParser concurrently with xml_init initialization ?
    If so I do not see how a function static object would prevent that.

On 09/22/2015 06:01 PM, David Wagner wrote:
>
> In xmlserializer/XmlUtil.cpp 
> <https://github.com/01org/parameter-framework/pull/246#discussion_r40105267>:
>
> > +#include <libxml/parser.h>
> > +
> > +/* Provides miscellaneous utility functions */
> > +class CXmlUtil
> > +{
> > +public:
> > +    /* xmlInitParser must be called in the main thread if multithreading is required
> > +     * @see http://xmlsoft.org/threads.html
> > +     * @see http://www.programmershare.com/1669782/
> > +     */
> > +    CXmlUtil() {
> > +        xmlInitParser();
> > +    }
> > +};
> > +
> > +static CXmlUtil xml_init;
>
> This introduced a potential race condition. In your second link, and 
> by reading libxml2's code, I understood that |xmlnitParser| can be 
> safely called multiple times, as long as it is called /sequentially/ 
> (as opposed to /in parallel/).
>
> It would be safer to put a static CXmlUtil object in a function (e.g. 
> in xmlserializer and call that function from ParameterMgr's 
> constructor) (See "effective C++": " C++'s guarantees that local 
> static objects are initialized when the object's definition is first 
> encountered during a call to that function." - that is not the case 
> for non-local objects, e.g. this |xml_init| object).
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/01org/parameter-framework/pull/246/files#r40105267>.
>

---------------------------------------------------------------------
Intel Corporation SAS (French simplified joint stock company)
Registered headquarters: "Les Montalets"- 2, rue de Paris, 
92196 Meudon Cedex, France
Registration Number:  302 456 199 R.C.S. NANTERRE
Capital: 4,572,000 Euros

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.